### DISCUSSION TITLE
[serializer.encode] - How to calculate the required size of the input array?

### DISCUSSION THREAD

**Discussion by EbiPenMan** [activej/activej#178] [Q&A]
How do you use the encoder? 
How much do you measure the byte array? 
How do you calculate the output byte size?

Couldn't you use the List<byte> instead of the byte[]?

**Comment by gkopff** [comment]
If I can also piggy-back on this ... assuming we've over-sized the buffer, how can we determine the actual number of bytes used?

Secondly, is there no mechanism to use an `OutputStream` so that the buffer size does not need to be pre-allocated?

Finally, do you have any documentation around the approximate serialisation sizes of common types?

**Comment by gkopff** [comment]
> Secondly, is there no mechanism to use an OutputStream so that the buffer size does not need to be pre-allocated?

Okay, I found a way of streaming ...

> If I can also piggy-back on this ... assuming we've over-sized the buffer, how can we determine the actual number of bytes used?


Assuming we're starting with a fresh `StreamOutput` each time we do this, is using `stream.pos()` the "right" way to do this?  Is there a better way than what's shown below:

```
    final StreamOutput stream = StreamOutput.create(new ByteArrayOutputStream());
    serializer.encode(stream.getBinaryOutput(), theObjectToSerialise);

    final int byteSize = stream.pos()
```

**Comment by eduard-vasinskyi** [comment]
Hello, @EbiPenMan, @gkopff

> 
> How do you use the encoder?
> How much do you measure the byte array?

You do not need to know the size of serialized data in advance. Just use a byte array that is big enough. You can look at the [ChannelSerializer](https://github.com/activej/activej/blob/master/core-datastream/src/main/java/io/activej/datastream/csp/ChannelSerializer.java), [StreamOutput](https://github.com/activej/activej/blob/master/core-serializer/src/main/java/io/activej/serializer/stream/StreamOutput.java) classes or [RedisConnection](https://github.com/activej/activej/blob/master/extra/cloud-redis/src/main/java/io/activej/redis/RedisConnection.java)’s `cmd()` method for an example. Although the classes are a bit complicated, the main idea is that you create a byte array of some initial size (whatever you like), try to encode data into the array, and catch `ArrayIndexOutOfBoundException`. Once, you catch it, it means that your array is not big enough for the data you try to serialize. Just create a bigger array and continue until serialization is successful. 

>
> How do you calculate the output byte size?
> If I can also piggy-back on this ... assuming we've over-sized the buffer, how can we determine the actual number of bytes used?

If you want to know the size of already serialized data, [BinarySerializer](https://github.com/activej/activej/blob/master/core-serializer/src/main/java/io/activej/serializer/BinarySerializer.java)’s `encode(byte[] array, int pos, T item)` actually returns a new position from which you can write to the array. So if you had called `encode(array, 10, item)` and it returned 14, that means 4 bytes had been written to the array. As for `encode(BinaryOutput out, T item)`, you can look up the difference between `out.pos()` after and before the serialization, that is the size of serialized data.

>
>Couldn't you use the List instead of the byte[]?

A `List<Byte>` would negate all the performance of serialization. It is magnitudes slower than operating directly on a byte array.
